source/common/network/connection_impl.h

```diff
@@ -167,16 +169,32 @@ class ConnectionImpl : public virtual Connection,
   // Callback issued when a delayed close timeout triggers.
   void onDelayedCloseTimeout();
 
+  void initializeDelayedCloseTimer();
+  bool inDelayedClose() const { return delayed_close_state_ != DelayedCloseState::None; }
+
   static std::atomic<uint64_t> next_global_id_;
 
+  // States associated with delayed closing of the connection (i.e., when the underlying socket is
+  // not immediately close()d as a result of a ConnectionImpl::close()).
+  enum class DelayedCloseState {
+    None,
+    // The socket will be closed immediately after the buffer is flushed _or_ if a period of
+    // inactivity after the last write event greater than or equal to delayed_close_timeout_ has
+    // elapsed.
+    CloseAfterFlush,
+    // The socket will be closed after a grace period of delayed_close_timeout_ has elapsed after
+    // the socket is flushed _or_ if a period of inactivity after the last write event greater than
+    // or equal to delayed_close_timeout_ has elapsed.
+    CloseAfterFlushAndTimeout
```

Comment:
nit: WDYT about calling this `CloseAfterFlushAndDelay`? Both of these states involve timeouts so IMO this is a little confusing.

Reply:
I considered `CloseAfterFlushAndDelay` but ultimately decided against it because both `CloseAfterFlush` and `CloseAfterFlushAndTimeout` introduce a delay into `close()` processing by waiting for either the flush or the <flush+timer trigger> to happen.

I could rename to `CloseAfterFlushAndWait` if you think that's clearer than `CloseAfterFlushAndTimeout`.

Reply:
Yeah I do think `Wait` is more clear than `Timeout` in this case if you don't mind.

Reply:
Done.